core/src/avm1/globals/xml_socket.rs

```diff
@@ -0,0 +1,267 @@
+use crate::avm1::function::FunctionObject;
+use crate::avm1::object::{NativeObject, Object};
+use crate::avm1::property_decl::define_properties_on;
+use crate::avm1::{property_decl::Declaration, ScriptObject};
+use crate::avm1::{Activation, Error, Executable, ExecutionReason, TObject, Value};
+use crate::context::{GcContext, UpdateContext};
+use crate::display_object::TDisplayObject;
+use crate::socket::SocketHandle;
+use crate::string::AvmString;
+use gc_arena::{Collect, Gc};
+use std::cell::{Cell, RefCell, RefMut};
+
+#[derive(Clone, Debug, Collect)]
+#[collect(require_static)]
+struct XmlSocketData {
+    handle: Cell<Option<SocketHandle>>,
+    /// Connection timeout in milliseconds.
+    timeout: Cell<u32>,
+    read_buffer: RefCell<Vec<u8>>,
+}
+
+#[derive(Clone, Debug, Collect)]
+#[collect(no_drop)]
+pub struct XmlSocket<'gc>(Gc<'gc, XmlSocketData>);
+
+impl<'gc> XmlSocket<'gc> {
+    pub fn handle(&self) -> Option<SocketHandle> {
+        self.0.handle.get()
+    }
+
+    pub fn set_handle(&self, handle: SocketHandle) -> Option<SocketHandle> {
+        self.0.handle.replace(Some(handle))
+    }
+
+    pub fn timeout(&self) -> u32 {
+        self.0.timeout.get()
+    }
+
+    pub fn read_buffer(&self) -> RefMut<'_, Vec<u8>> {
+        self.0.read_buffer.borrow_mut()
+    }
+
+    pub fn cast(value: Value<'gc>) -> Option<Self> {
+        if let Value::Object(object) = value {
+            if let NativeObject::XmlSocket(xml_socket) = object.native() {
+                return Some(xml_socket);
+            }
+        }
+        None
+    }
+}
+
+const PROTO_DECLS: &[Declaration] = declare_properties! {
+    "timeout" => property(get_timeout, set_timeout);
+    "close" => method(close);
+    "connect" => method(connect);
+    "send" => method(send);
+    "onConnect" => method(on_connect; DONT_ENUM | DONT_DELETE);
+    "onClose" => method(on_close; DONT_ENUM | DONT_DELETE);
+    "onData" => method(on_data; DONT_ENUM | DONT_DELETE);
+    "onXML" => method(on_xml; DONT_ENUM | DONT_DELETE);
+};
+
+fn get_timeout<'gc>(
+    _activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(xml_socket) = XmlSocket::cast(this.into()) {
+        Ok(xml_socket.timeout().into())
+    } else {
+        Ok(Value::Undefined)
+    }
+}
+
+fn set_timeout<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(xml_socket) = XmlSocket::cast(this.into()) {
+        let timeout = args
+            .get(0)
+            .unwrap_or(&Value::Undefined)
+            .coerce_to_u32(activation)?;
+
+        // FIXME: Check if flash player clamps this to 250 like AS3 sockets.
+        xml_socket.0.timeout.set(timeout);
```

Comment:
Consistency: there's a `socket.set_timeout()` in Avm2 code and a `timeout()` on `XmlSocket` just above, maybe a `xml_socket.set_timeout()` makes sens? By the way the clamping could be added in this function in the future if needed :slightly_smiling_face: 

```suggestion
        xml_socket.set_timeout(timeout);
```